pkgs/dart_tooling_mcp_server/lib/src/utils/json.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Utility for indexing json data structures.
+///
+/// Each element of [path] should be a `String`, `int` or `(String, String)`.
+///
+/// For each element `key` of [path], recurse into [json].
+///
+/// If the `key` is a String, the next json structure should be a Map, and have
+/// `key` as a property. Recurse into that property.
+///
+/// If `key` is an `int`, the next json structure must be a List, with that
+/// index. Recurse into that index.
+///
+/// If `key` is a `(String k, String v)` the next json structure must be a List
+/// of maps, one of them having the property `k` with value `v`, recurse into
+/// that map.
+///
+/// If at some point the types don't match throw a [FormatException].
+///
+/// Returns the result as a [T].
+T dig<T>(dynamic json, List<dynamic> path) {
```

Comment:
```suggestion
T dig<T>(dynamic json, List<Object> path) {
```

It looks like the only value that can result from a `null` path entry is a `null` value. Do we need that behavior? Can we type this with non-null path values?